### ISSUE TITLE
ActiveForm inputs with custom id are not validated

### ISSUE THREAD

**Issue by SDKiller** [yiisoft/yii2#7627]
If you pass your custom `id` like this:

```php
$form->field($model, 'comment')->textarea(['id' => 'myCustomId'])
```

I suppose the reason is in `\yii\widgets\ActiveField::getClientOptions`:

```php
        $inputID = Html::getInputId($this->model, $this->attribute);
        $options['id'] = $inputID;
```

It does not care if `id` was passed to field options.

I cannot use autogenerated `id` as there are several forms of the same model on the page - `ids` will not be unique.

**Comment by SDKiller** [comment]
UPD

I see this in docs
```
* If you set a custom `id` for the container element, you may need to adjust the [[$selectors]] accordingly.
```

But what is the sense to autogenerate `id` for input if this `id` does not exist in `DOM` as custom id was set explicitly?

**Comment by cebe** [comment]
is it clear now, or is the question still open?

**Comment by SDKiller** [comment]
As I could understand - `id` is used internally and not always represents `id` of html element `.
While it's confusing and  inconvenient to specify in addition to `id` of html element 

```php
$form->field($model, 'comment', ['selectors' => ['input' => '#myCustomId']])
            ->textarea(['id' => 'myCustomId']);?>
```

**Comment by cebe** [comment]
as far as I see, yii should be able to automatically update the selector in most cases. Need to check that.

**Comment by SDKiller** [comment]
In my example `options` passed to `textarea()` have no chance to be fetched by `getClientOptions()` as they are merged with `inputOptions` only for `$this->parts['{input}']` rendering.

They are not merged into `inputOptions` property of `ActiveField` object itself.

**Comment by SDKiller** [comment]
I mean in `\yii\widgets\ActiveField::getClientOptions` it could be done like this

```php
$inputID = isset($this->inputOptions['id']) ?  $this->inputOptions['id'] : Html::getInputId($this->model, $this->attribute);
```

but it will work only if you create field in this way:

```php
$form->field($model, 'comment', ['inputOptions' => ['id' => 'myCustomId']])->textarea();?>
```

**Comment by dynasource** [comment]
I can confirm this issue and the solution of SDKiller to be working for my scenario. Can it be implemented? thnx

**Comment by cebe** [comment]
seems there is no good way to solve it.

**Comment by dynasource** [comment]
I can confirm this issue and the solution provided by SDKiller to be working for me.
Two lines should be replaced:

* https://github.com/yiisoft/yii2/blob/master/framework/widgets/ActiveField.php#L216
* https://github.com/yiisoft/yii2/blob/master/framework/widgets/ActiveField.php#L715

**Comment by cebe** [comment]
I see, so this would fix it for at least some of the cases, right? can you send a pull request?

**Comment by dynasource** [comment]
I will create one now

**Comment by dynasource** [comment]
@cebe: is it possible to fix this one on short term?

**Comment by dizeee** [comment]
The solution works with client validation but if `ActiveForm::$enableClientValidation` is set to `false` and `ActiveForm::$enableAjaxValidation` is set to `true`, than validation doesn't work no more because errors returned from the ajax validation are still formatted as `{"model-attribute": ["Error"]}`, but the attribute id in clientOptions is now set to `customid`, not `model-attribute`, so the script doesn't see the attribute errors in the response. I think the proper code in \yii\widgets\ActiveField::getClientOptions() should be:
```
$inputID = $this->getInputId();
$options['id'] = Html::getInputId($this->model, $this->attribute);
// not $options['id'] = $inputID
```
I'll tinker a bit with it and make a pull request if I'll find this working for both client and ajax validation.

**Comment by bSushil** [comment]
Form validation doesn't work when you give custom id to the form. I am using Yii 2.0.13. I would prefer if validation for form works even if we provide custom id to the form.

**Comment by cebe** [comment]
are you sure the id is unique? otherwise there should be no problem adding an id to the form. Please open a new issue with more detail if the problem persists.

**Comment by themartianmw** [comment]
So, there's no solution for the problem mentioned by SDKiller? I need to work with custom IDs too, but validation doesn't work with them.

**Comment by cebe** [comment]
@thehereward this issue is fixed since 2.0.7, if you have an issue with the latest version, please open a new issue.

**Comment by themartianmw** [comment]
@cebe your solution (in 2.0.7) was discarded in 2.0.8 because of the issue 7248. However I found a better way to do this, watching the code of yii2-dynamic-form from **2amigos** (specifically the code for the view). And this is enough for me because I only needed the custom IDs for a dynamic form.
